File: apps/rapid-mac/Sources/Rapid/Chat/ConversationSearch.swift
import Foundation

/// Pure conversation-search and grouping logic shared by the search surface
/// and its tests. Search stays local: it only inspects the already-loaded
/// history snapshot and never sends transcript content anywhere.
enum ConversationSearch {
    struct Section {
        enum Bucket: Hashable {
            case today
            case yesterday
            case previous7Days
            case previous30Days
            case older
        }

        let bucket: Bucket
        let conversations: [ChatConversation]
    }

    /// Match every whitespace-delimited query term against the conversation's
    /// title or visible user/assistant prose. Terms may land in different
    /// messages, which makes a query such as "swift cache" useful even when
    /// the two words came from different turns.
    static func results(
        in conversations: [ChatConversation],
        matching query: String
    ) -> [ChatConversation] {
        let terms = searchTerms(from: query)
        return conversations
            .filter { conversation in
                guard !terms.isEmpty else { return true }
                let searchableText = [conversation.title] + conversation.messages.compactMap { message in
                    switch message.role {
                    case .user, .assistant:
                        return message.content
                    case .system, .tool, .unknown:
                        return nil
                    }
                }
                return terms.allSatisfy { term in
                    searchableText.contains { text in
                        text.localizedStandardContains(term)
                    }
                }
            }
            .sorted { $0.updatedAt > $1.updatedAt }
    }

    /// Date groups used by the search panel. Archived conversations remain in
    /// these groups because search is their direct recovery path too.
    static func sections(
        for conversations: [ChatConversation],
        now: Date,
        calendar: Calendar = .current
    ) -> [Section] {
        let sorted = conversations.sorted { $0.updatedAt > $1.updatedAt }
        let startOfToday = calendar.startOfDay(for: now)
        let startOfYesterday = calendar.date(byAdding: .day, value: -1, to: startOfToday)
        let weekCutoff = calendar.date(byAdding: .day, value: -7, to: startOfToday)
        let monthCutoff = calendar.date(byAdding: .day, value: -30, to: startOfToday)

        var today: [ChatConversation] = []
        var yesterday: [ChatConversation] = []
        var week: [ChatConversation] = []
        var month: [ChatConversation] = []
        var older: [ChatConversation] = []

        for conversation in sorted {
            if calendar.isDate(conversation.updatedAt, inSameDayAs: now) {
                today.append(conversation)
            } else if let startOfYesterday,
                      conversation.updatedAt >= startOfYesterday,
                      conversation.updatedAt < startOfToday {
                yesterday.append(conversation)
            } else if let weekCutoff, conversation.updatedAt >= weekCutoff {
                week.append(conversation)
            } else if let monthCutoff, conversation.updatedAt >= monthCutoff {
                month.append(conversation)
            } else {
                older.append(conversation)
            }
        }

        return [
            Section(bucket: .today, conversations: today),
            Section(bucket: .yesterday, conversations: yesterday),
            Section(bucket: .previous7Days, conversations: week),
            Section(bucket: .previous30Days, conversations: month),
            Section(bucket: .older, conversations: older),
        ].filter { !$0.conversations.isEmpty }
    }

    private static func searchTerms(from query: String) -> [String] {
        query
            .split(whereSeparator: { $0.isWhitespace })
            .map(String.init)
    }
}

File: apps/rapid-web/frontend/src/chat/ConversationSearch.ts
import type { Conversation } from '@/state/types';
import { dateGroupOf, type DateGroup } from '@/lib/format';

/**
 * Conversation search and grouping.
 *
 * Ported from `apps/rapid-mac/Sources/Rapid/Chat/ConversationSearch.swift`,
 * including the two decisions that are easy to get wrong:
 *
 *   - **All terms, anywhere.** Each whitespace-delimited term matches the
 *     title OR any message, and they may land in different messages — which is
 *     what makes "swift cache" useful across two turns.
 *   - **Every branch, not just the visible path.** An answer you regenerated
 *     away is exactly what you come back looking for.
 *
 * Search stays local: it reads the already-loaded history and sends nothing.
 */

const GROUP_ORDER: DateGroup[] = [
  'Today',
  'Yesterday',
  'Previous 7 days',
  'Previous 30 days',
  'Older',
];

export interface SearchSection {
  label: string;
  rows: Conversation[];
}

/** Split a query into terms, dropping the empties an extra space produces. */
export function searchTerms(query: string): string[] {
  return query
    .trim()
    .toLowerCase()
    .split(/\s+/)
    .filter((term) => term !== '');
}

/** Conversations matching every term, most recently updated first. */
export function searchConversations(
  conversations: Conversation[],
  query: string,
): Conversation[] {
  const terms = searchTerms(query);
  const matched =
    terms.length === 0
      ? [...conversations]
      : conversations.filter((conversation) => {
          const haystack = [
            conversation.title,
            ...conversation.nodes.map((node) => node.content),
          ]
            .join('\n')
            .toLowerCase();
          return terms.every((term) => haystack.includes(term));
        });
  return byRecency(matched);
}

/**
 * Group into date sections, pinned first.
 *
 * Archived conversations are NOT filtered out here — the Mac app takes the
 * same position, because search is an archived conversation's direct recovery
 * path. Callers that show a plain list filter before calling.
 */
export function groupConversations(
  conversations: Conversation[],
  now: number,
): SearchSection[] {
  const pinned = conversations.filter((conversation) => conversation.isPinned);
  const rest = conversations.filter((conversation) => !conversation.isPinned);

  const buckets = new Map<DateGroup, Conversation[]>();
  for (const conversation of rest) {
    const group = dateGroupOf(conversation.updatedAt, now);
    const list = buckets.get(group) ?? [];
    list.push(conversation);
    buckets.set(group, list);
  }

  const sections: SearchSection[] = [];
  if (pinned.length > 0) sections.push({ label: 'Pinned', rows: byRecency(pinned) });
  for (const group of GROUP_ORDER) {
    const rows = buckets.get(group);
    if (rows?.length) sections.push({ label: group, rows: byRecency(rows) });
  }
  return sections;
}

export function byRecency(rows: Conversation[]): Conversation[] {
  return [...rows].sort((a, b) => b.updatedAt - a.updatedAt);
}
